### ISSUE TITLE
Dynamic Scaling: Adding new partitions

### ISSUE THREAD

**Issue by lenaschoenburg** [camunda/camunda#21439]
## Goal
After making the number of brokers and replication factor dynamic, the next step is to allow _adding_ new partitions to scale up.

Scaling up partitions will be seamless for users. Users don't need to re-deploy resources, adjust their processes or their API usage.
The performance impact  during scaling should be as small as possible, we try not to interrupt regular processing too much and keep memory and state size overhead as low as possible.

Scaling down by _removing_ partitions is not in scope, scaling down is only possible by removing brokers.

This implements [PDP 2226](https://github.com/camunda/product-hub/issues/2226).

## Prerequisites

These are features and improvements that we anticipate to become necessary for the implementation of milestone 1 and 2.
They can be implemented first and don't have any direct dependency on scaling.

```[tasklist]
### Improved command distribution
- [ ] Ordered command distribution
- [ ] Command distribution can write a custom continuation command
- [ ] Command distribution is observable
```

## Milestone 1

The first deliverable milestone will be when newly added partitions are usable and participate in most processing. Messages will only be published on the old partitions and accordingly, message subscriptions are also created on old partitions only. This slightly limits the effectiveness of scaling up because message related processing might still be bottlenecked by the old partition count. Nevertheless, the first milestone is usable and beneficial, especially for workloads that don't rely much on messages.

### Acceptance Criteria

- [ ] Resources are available on all partitions
- [ ] Backups are created and restored with all partitions
- [ ] Process instances are created on all partitions
- [ ] Processing is working as usual, all features are usable
- [ ] Performance impact of scaling is minimal, no interruptions

```[tasklist]
### Dynamic routing info
- [ ] Add a draft title or issue reference here
```

```[tasklist]
### Starting new partitions
- [ ] Add a draft title or issue reference here
```

```[tasklist]
### Re-distributing resources
- [ ] Add a draft title or issue reference here
```

```[tasklist]
### Observable scaling
- [ ] Metrics for resource re-distribution
- [ ] Cluster status query includes progress indicators
```

## Milestone 2

The second milestone improves scalability further by including new partitions in message processing. Existing messages and subscriptions will be relocated to new partitions, ensuring an even distribution message processing across all partition.

### Acceptance Criteria

- [ ] Messages are published on all partitions
- [ ] Existing message subscriptions are relocated as needed
- [ ] [Guarantees around messages](https://docs.camunda.io/docs/components/concepts/messages), in particular cardinality, uniqueness, ordering etc. are maintained
- [ ] Message correlation delay during scaling is kept at a minimum

```[tasklist]
### Relocating messages
- [ ] Add a draft title or issue reference here
```

## Documentation

Although the feature is fairly complex, the documentation effort is mostly internal. Perhaps the most important aspect for user documentation is clarifying that scaling down is not possible.

```[tasklist]
### External documentation
- [ ] Add a new partition scaling page 
- [ ] API reference is extended with new endpoints
```

```[tasklist]
### Internal documentation
- [ ] Chaos experiments to demonstrate partition scaling
- [ ] Developer documentation for message relocation
- [ ] Developer documentation for resource redistribution
```

**Comment by lenaschoenburg** [comment]
## Redistributing Resources

### Challenges:
- Many different resource types
- Too many resources to redistribute them all at once
- New resources can be created while redistributing
- Resources can be modified while redistributing
- Resources can be deleted while redistributing

### Proof of Concept Solution

For each resource type, enqueue batches of resources until all resources of that type are redistributed and then move on to the next resource type.

New resources deployed during redistribution might get distributed twice because command distribution immediately starts distributing to all partitions.

Resource deletion is handled as a special case, deleting a resource already redistributed will also distribute the deletion while deleting a resource not yet redistributed will only delete locally.
#### Problems

This approach cannot deal with updating resources while redistributing. This would require additional special handling of updates, similar to resource deletion.

The handling of resource deletion and updates will leak into various processors which make long-term maintenance difficult.

Duplicated resource distribution is inefficient and might cause at least intermittent errors.

Immediate distribution of new resources to partitions that have not yet received old resources might cause at least intermittent errors, for example because a role is created for a user that isn't available yet.

### Proposed Solution

First ensure that all resources are distributed in-order, with a queue per resource type (or several related resource type as done for users, roles etc.). Notably, this would introduce ordering for deploying processes. The same queue is used for resource deletion and updates.

Then introduce a new `RedistributionState` that keeps track of `(partitionId, resourceType) -> maxRedistributedResourceKey`. We update this state every time we enqueue a new batch of resources for redistribution.

Extend the command distribution behavior with `forPartitionsWithResource(ResourceType resourceType, long resourceKey)`. The behavior uses the `RedistributionState` to decide which partitions are ready to receive the command. If a partition does not know about a resource yet, don't distribute the command to that partition, it will get created/updated/deleted locally and the correct state will be enqueued for redistribution later anyway. If the partition already knows about that resource, distribute the new command for distribution to that partition.

This is a generalization of the special handling of resource deletion described in the POC.

#### FAQ

- Why require ordered distribution?
	- So that we can ensure that initial creation of the resource through redistribution happens before resource deletion/update.
- Can't we just enqueue every new resource deletion/update distribution for all partitions immediately?
	- No, because the redistribution of resources does not happen at once. Instead, we enqueue them in batches. Until the resources creation is enqueued, a resource update both modifies the resource locally and enqueues the command for distribution. The enqueued update is no longer valid because the creation command will already reflect the updated state.

**Comment by remcowesterhoud** [comment]
@lenaschoenburg I thought about this problem a lot over the weekend, and to be honest, I struggle to come up with a solution.

Whilst your proposed solution seems good ast first, I think there's cases where it won't work. For example, when we assign a Role to a User. In this scenario we need to make sure both the Role and the User are created on the new partition, before we can distribute the assigning itself. This is also the reason why we needed an Identity distribution queue, as opposed to more atomic ones, such as User, Role, etc.

Another exception I can think of is resource deletion. We deploy resources on a Deployment level (multiple resources), but we delete them on a single resource level. I'd assume redistribution would also work on a single resource level, so I don't see how you could compare those keys.

**Comment by lenaschoenburg** [comment]
> Another exception I can think of is resource deletion

I think this is covered nicely now: While redistributing resources, we keep track of the highest resource key redistributed to each partition. When deleting resources, we check whether a partition already received the resource to decide whether we need to distribute the deletion or not.

> I think there's cases where it won't work. For example, when we assign a Role to a User

I think this should also work: When adding an entity to a role, we check whether a partition has already received the role or not to decide whether we need to distribute the command already or not.
I'm not super familiar with the new User/Role handling though so perhaps I'm missing something here @remcowesterhoud.

**Comment by remcowesterhoud** [comment]
>I think this should also work: When adding an entity to a role, we check whether a partition has already received the role or not to decide whether we need to distribute the command already or not. Users are redistributed before roles, so these already exist.

What I struggle with is how redistribution of these Role-User relationships works. Let's say I have this in my state:

```
Role X
User Y - Assigned to Role X
```

On scaling this redistribution happens:

```
Create User Y
Create Role X
Assign User Y to Role X
```

What if during the scaling Partition 1 receives a command to remover User Y from Role X, but this happens before we've redistributed the assigning of the role? So:

```
Redistribution batch 1
- Redistribute Create User Y
- Redistribute Create Role X

Regular processing:
- Remove User Y from Role X

Redistribution batch 2
- Redistribute Assign User Y to Role X
```

How can we determine that we shouldn't distribute `Remove User Y from Role X` yet, but we need to queue it until the 2nd redistribution batch completed?

**Comment by lenaschoenburg** [comment]
I see, thanks for the explanation :+1: 

I think the problem stems from the fact that the _assignment_ itself is not represented as a resource itself but only exists as properties of the entity (i.e. user) and role.
We could either create the assignment too early, when creating the user, or later, when creating the role. With the latter, we could rely on the existence of the role to check whether we can distribute the assignment or not.

**Comment by npepinpe** [comment]
As I'm the one least familiar with the engine, I won't go too much in details, and please correct me if I make some wrong assumptions about internals.

- When we start the redistribution, we can keep track of the highest resource key for all resource as a threshold. Until a partition has seen _all_ of these (so it has seen all resources previous distributed before the start of the scaling operation), it is _inactive_ (or whatever you want to call it).
- Any distributed command is to an inactive partition is buffered until the partition is active. This _may_ cause unnecessary work (e.g. deploy a process only to remove it later), but considering deployments/deletions are typically not the "hot" path, I think this is likely OK.
  - We should probably add some metrics to monitor size of buffered command distribution just in case.
  - There is an edge case where a resource is deleted locally and the command buffered for distribution, so we will eventually distributed the delete command but will never have distributed the resource itself. I think that's something we can likely handle (though not sure in a nice way).

I think the biggest issue is handling the resource dependencies. I can also imagine this causing issues in the future as we add new resources :(

**Comment by lenaschoenburg** [comment]
I discussed the problem of resource redistribution with @npepinpe and @remcowesterhoud.
We think the approach should work in general if we keep track of _composite keys_ for virtual resources such as role assignments and authorizations.

Something that is not yet entirely clear is how to deal with deployments. One problem is that multiple resource types can get deployed together. This makes checking of redistribution progress difficult and may require us to split up the deployment before distributing. Another problem is version (tag) binding. Depending on the implementation and semantics, we might need to redistribute resources that are bound to each other together.

Since the identity resources are not finalized yet, I'll start with processes and deployments. Hopefully we can validate the overall solution already.
